### PR TITLE
8277459: Add jwebserver tool

### PR DESCRIPTION
**Pull request by FrauBoes**
This change introduces jwebserver, a dedicated JDK tool for the Simple Web Server. 

A description is provided in a follow-up comment.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8277459](https://bugs.openjdk.java.net/browse/JDK-8277459): Add jwebserver tool

### Reviewers
 * [Michael McMahon](https://openjdk.java.net/census#michaelm) (@Michael-Mc-Mahon - **Reviewer**) ⚠️ Review applies to 60acea9d698ccf4cb788b41cc12abdcc4edcaa98
 * [Daniel Fuchs](https://openjdk.java.net/census#dfuchs) (@dfuch - **Reviewer**) ⚠️ Review applies to 81499efab84be3abdbfe8a21494044f25588423d
 * [Magnus Ihse Bursie](https://openjdk.java.net/census#ihse) (@magicus - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6497/head:pull/6497` \
`$ git checkout pull/6497`

Update a local copy of the PR: \
`$ git checkout pull/6497` \
`$ git pull https://git.openjdk.java.net/jdk pull/6497/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6497`

View PR using the GUI difftool: \
`$ git pr show -t 6497`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6497.diff">https://git.openjdk.java.net/jdk/pull/6497.diff</a>

</details>

### CODE REVIEW

**Comment by FrauBoes** [comment]
This change adds a dedicated command-line tool for the Simple Web Server, to improve convenience and approachability. The new tool `jwebserver` calls the main entry point of the jdk.httpserver module, thus hiding this implementation detail which is not hidden with `java -m jdk.httpserver`. The command-line options are identical, the only addition is support for `-version` / `--version`. 

A man page for `jwebserver` is added and the module and package level documentation is updated to reflect the new tool.

A new internal class JWebServer is added to be able to distinguish whether the server was launched via `jwebserver` or `java -m jdk.httpserver`.

The new tests are copies of the existing command-line tests.

CSR: https://bugs.openjdk.java.net/browse/JDK-8277460

File: src/jdk.httpserver/share/classes/module-info.java [hunk 1]
```diff
@@ -24,41 +24,21 @@
...
- *    -h, -?, --help        - Print this help message.
- * }</pre>
+ * Defines the JDK-specific HTTP server API, and provides the jwebserver tool
+ * for running a minimal HTTP server.
+ *
+ * <p>The com.sun.net.httpserver package defines a high-level API for building
+ * servers that support HTTP and HTTPS. The SimpleFileServer class implements a
+ * simple HTTP-only file server intended for testing, development and debugging
+ * purposes. A default implementation is provided via the {@code jwebserver} tool
+ * and the main entry point of the module, which can be invoked with
```
Comment (dfuch) [inline src/jdk.httpserver/share/classes/module-info.java:34]: should we say "which can also be invoked ..." ?

File: src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/SimpleFileServerImpl.java [hunk 2]
```diff
@@ -71,11 +71,12 @@
      *
      * @param  writer the writer to which output should be written
      * @param  args the command line options
+     * @param launcher the launcher the server is started from
```
Comment (dfuch) [inline src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/SimpleFileServerImpl.java:74]: should this be aligned like the other `@param` tags?

File: src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/resources/simpleserver.properties [hunk 3]
```diff
@@ -23,10 +23,17 @@
 # questions.
 #
 
-usage=\
+usage.java=\
 Usage: java -m jdk.httpserver [-b bind address] [-p port] [-d directory]\n\
 \                              [-o none|info|verbose] [-h to show options]
 
+usage.jwebserver=\
```
Comment (dfuch) [inline src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/resources/simpleserver.properties:30]: should usage also show `-version, --version` in both usage strings, and in the usage summary of the man page below as well?

File: src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/resources/simpleserver.properties [hunk 4]
```diff
@@ -35,6 +42,7 @@ Options:\n\
 -o, --output          - Output format. none|info|verbose. Default: info.\n\
 -p, --port            - Port to listen on. Default: 8000.\n\
 -h, -?, --help        - Print this help message.\n\
+-version, --version   - Print version information.\n\
```
Comment (dfuch) [inline src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/resources/simpleserver.properties:45]: Should this be " - Print version information and exit."

File: src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/JWebServer.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     * or an I/O error occurs, the server is not started and this method invokes
+     * System::exit with an appropriate exit code.
+     *
+     * @param args the command-line options
+     * @throws NullPointerException if {@code args} is {@code null}, or if there
+     *         are any {@code null} values in the {@code args} array
+     */
+    public static void main(String... args) {
+        int ec = SimpleFileServerImpl.start(new PrintWriter(System.out, true, UTF_8), "jwebserver", args);
+        if (ec != 0)
```
Comment (magicus) [inline src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/JWebServer.java:60]: Personally I strongly dislike multiline if-statements without braces. They are an invitation to mistakes. Please reconsider putting this on one line, or adding braces.

**Comment by magicus** [comment]
Build changes look good.

File: src/jdk.httpserver/share/classes/com/sun/net/httpserver/SimpleFileServer.java [hunk 6]
```diff
@@ -104,12 +104,10 @@
...
  *
- * <h2>Main entry point</h2>
+ * <h2>jwebserver Tool</h2>
  *
- * <p>A <a id="server-impl">simple HTTP file server implementation</a> is
- * provided via the
- * <a href="{@docRoot}/jdk.httpserver/module-summary.html#entry-point">main entry point</a>
- * of the {@code jdk.httpserver} module.
+ * <p>A simple HTTP file server implementation is provided via the
+ * {@code jwebserver} tool.
```
[the remark this answers is not in the document]
Comment (FrauBoes) [reply src/jdk.httpserver/share/classes/com/sun/net/httpserver/SimpleFileServer.java:110]: The tag can currently only be added to module and package summaries, if I read the code correctly, see `open/make/jdk/src/classes/build/tools/taglet/ToolGuide.java`. I can check if we can add class-level support at a later point.

File: src/jdk.httpserver/share/classes/module-info.java [hunk 7]
```diff
@@ -24,41 +24,21 @@
...
+ * <p>The {@link com.sun.net.httpserver} package defines a high-level API for
+ * building servers that support HTTP and HTTPS. The SimpleFileServer class
+ * implements a simple HTTP-only file server intended for testing, development
+ * and debugging purposes. A default implementation is provided via the
+ * {@code jwebserver} tool and the main entry point of the module, which can
+ * also be invoked with {@code java -m jdk.httpserver}.
+ *
+ * <p>The {@link com.sun.net.httpserver.spi} package specifies a Service Provider
+ * Interface (SPI) for locating HTTP server implementations based on the
+ * com.sun.net.httpserver API.
```
Comment (dfuch) [inline src/jdk.httpserver/share/classes/module-info.java:39]: Nit: maybe this should be ` * {@code com.sun.net.httpserver} API.`

**Review by Michael-Mc-Mahon — APPROVED** [review]
LGTM

File: src/jdk.httpserver/share/classes/com/sun/net/httpserver/SimpleFileServer.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (FrauBoes) [reply src/jdk.httpserver/share/classes/com/sun/net/httpserver/SimpleFileServer.java:110]: Support for `@toolGuide` in class-level documentation was added to the mainline with https://github.com/openjdk/jdk/commit/e3e5908d0d385ef908ba9752908aaf28b4b3e4f4

File: src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/Main.java [hunk 8]
```diff
@@ -56,10 +56,10 @@
...
     public static void main(String... args) {
-        int ec = SimpleFileServerImpl.start(new PrintWriter(System.out, true, UTF_8), args);
-        if (ec != 0)
+        int ec = SimpleFileServerImpl.start(new PrintWriter(System.out, true, UTF_8), "java", args);
+        if (ec != 0) {
             System.exit(ec);
-        // otherwise the server has been started successfully and runs in
-        // another non-daemon thread.
+        }  // otherwise the server has been started successfully and runs in
+           // another non-daemon thread.
```
[the remark this answers is not in the document]
Comment (FrauBoes) [reply src/jdk.httpserver/share/classes/sun/net/httpserver/simpleserver/Main.java:63]: Right, good to mention those two cases.

**Review by dfuch — APPROVED** [review]
LGTM!

**Pull request closed without merging**